tests/test_bookmarks.py

```diff
@@ -0,0 +1,211 @@
+            if self.is_incremental(stream):
+                for key, value in bookmark_dict.items():
+                    # this checks if we replicated any record as start date has format '%Y-%m-%dT%H:%M:%SZ'
+                    try:
+                        second_start_date = (datetime.strptime(value, '%Y-%m-%dT%H:%M:%S.%fZ') - timedelta(days=1)).strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+                        new_state['bookmarks'][stream][key] = second_start_date
+                    except ValueError:
+                        new_state['bookmarks'][stream][key] = value
+
+        # Set state for next sync
+        menagerie.set_state(conn_id, new_state)
+
+        ##########################################################################
+        ### Second Sync
+        ##########################################################################
+
+        # Run a sync job using orchestrator
+        second_sync_record_count = self.run_and_verify_sync(conn_id)
+        second_sync_records = runner.get_records_from_target_output()
+        second_sync_bookmarks = menagerie.get_state(conn_id)
+
+        for stream in expected_streams:
+
+            with self.subTest(stream=stream):
+                # collect information for assertions from syncs 1 & 2 base on expected values
+                first_sync_count = first_sync_record_count.get(stream, 0)
+                second_sync_count = second_sync_record_count.get(stream, 0)
+
+                # collect information for assertions from syncs 1 & 2 base on expected values
+                first_sync_messages = [record.get('data') for record in first_sync_records.get(stream).get('messages')
+                                       if record.get('action') == 'upsert']
+                second_sync_messages = [record.get('data') for record in second_sync_records.get(stream).get('messages')
+                                        if record.get('action') == 'upsert']
+
+                first_bookmark_key_value = first_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)
+                second_bookmark_key_value = second_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)
+
+                if self.is_incremental(stream):
+
+                    self.assertEqual(first_bookmark_key_value, second_bookmark_key_value)
+
+                    # if the stream is child stream, then collect parent stream records for collecting ids
+                    if self.parent_child_mapping.get(stream):
+                        parent_records = [record.get('data') for record in first_sync_records.get(self.parent_child_mapping.get(stream) + 's').get('messages')
+                                        if record.get('action') == 'upsert']
+
+                    # NOTE: Sample bookmark:
+                    #   "ad_accounts": {
+                    #       "updated_at(parent_organization_id:organization-id-123)": "2020-01-01T00:00:00Z"
+                    #   }
```

Comment:
I'm not sure this is what the bookmark structure should look like. Here is how tap-google-ads handled top level accounts. https://github.com/singer-io/tap-google-ads/blob/5e019075074237b8553d40b1f057b2e2c5939cf2/tests/test_google_ads_bookmarks.py#L106-L108

Reply:
This Tap already has this bookmarking structure for parent-child INCREMENTAL streams

Reply:
The comment does not match the structure that I linked.

Reply:
The bookmark structure is already present in the existing Tap, as noted in the previous comment. So should we change that? If yes then should we change it as part of the current release or put it as a backlog card? Also, we have to handle backward compatibility there.